### DISCUSSION TITLE
What is the reasoning behind best practices of Relay?

### DISCUSSION THREAD

**Discussion by Jabher** [WalletConnect/walletconnect-monorepo#2396] [Developer Support] [answered]
We're moving to v2 and trying to understand the reasoning behind best practices, and how the granularity should work.
Should I use separate project ID for testnet domain? For every sub-app? What is the reasoning? 
Documentation is not giving this answer.

I'm specifically talking about https://docs.walletconnect.com/2.0/cloud/relay
> Best Practices
> Create a new projectId for each project. This allows for more granular control, dedicated explorer listings, and project metrics.
> **Don't reuse a projectId.** - why?
> Use the AllowList to limit unauthorized usage.
> Avoid committing projects keys to the repo. Use env variables instead.

**Comment by finessevanes** [comment]
Hey, asking our team for an explanation.

**Comment by chendinh** [comment]
Im not walletconnect team, however i have some ideas about this. 
actually v2 is not different from v1 so much.
i used to update my app from v1 to v2. 
- With v1, each connection/socket is managed by each instant walletconnect-client. With v2, you must create a HOOK or the component which cover all pages in your app. you use the only one instant walletconnect-client-v2 for everything, 
- This v2 instant will manage each connection by pairings and sessions. You should init instant once from beginning and subscribe sockets after your app open again ([Background and Foreground]
- paringID is important to reparated betweeen connections. sessions is important too, you can check if session is active or not to disconnect it or reconnect pairing.
This is some information which is not in document.


"Should I use separate project ID for testnet domain? For every sub-app? What is the reasoning?"
-> you dont need separate project ID for testnet domain 
-> you need separate project ID for sub-app ( if sub-app is different app for other purposes or other catetories)
The reason ? i guess, wallectconnect will have premium connection and the own token crypto. They wanna check how many requests from each app. by the other hand, they can block your project easily if it spam server or scam project

**Comment by finessevanes** [comment, marked as answer]
When migrating to v2, the decision to use separate project IDs for the testnet domain or each sub-app depends on metrics and allowlist separation requirements. If you need distinct metrics or separate allowlists, it is recommended to have separate projects. However, if those factors are not crucial, managing a single project is a more convenient option.
